### PR TITLE
Train the data with nn classifier and plot the results

### PR DESCRIPTION
**Pull request by suraz09**
- Created a basic nn classifier
- Trained the classifier
-  Plotted the results of the classifier

### CODE REVIEW

File: src/analysis/visualize.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import numpy as np
+np.random.seed(7)
+import matplotlib.pyplot as plt
+
+import seaborn as sns
```
Comment (Mitchellpkt) [inline src/analysis/visualize.py:2]: I prefer to lump all of the `import` functions toward the top, then make settings such as random seed and sns style closer to the relevant code. That's just a stylistic preference though, this is OK as is.

File: src/analysis/visualize.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+sns.set(style="white", palette="muted", color_codes=True, context="talk")
+
+create_gif = False
+
+"""
+    Show the results in the grahph plot.
```
Comment (Mitchellpkt) [inline src/analysis/visualize.py]: (typo)

Comment (Mitchellpkt) [reply src/analysis/visualize.py]: Maybe I started the review in the wrong place or missed some documentation. However at the moment I don't know what results this refers to?? I need some comments with context about the function & inputs.

File: src/analysis/visualize.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+def plot_distributions(y, Z, iteration=None, val_metrics=None, p_rules=None, fname=None):
+    fig, axes = plt.subplots(1, 1, figsize=(10, 4), sharey=True)
+    legend = {'race': ['African-American', 'Others']}
+    for idx, attr in enumerate(Z.columns):
+        for attr_val in [0, 1]:
+            ax = sns.distplot(y[Z[attr] == attr_val], hist=False,
+                              kde_kws={'shade': True, },
+                              label='{}'.format(legend[attr][attr_val]))
+        ax.set_xlim(0, 1)
+        ax.set_ylim(0, 7)
```
Comment (Mitchellpkt) [inline src/analysis/visualize.py:26]: Why is the `7` hard-coded in? What does it represent?

Comment (Mitchellpkt) [reply src/analysis/visualize.py:26]: In general, preferable to have a function like `ax.set_ylim(0, max(y_data))` or something like that.

File: src/analysis/visualize.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            ax.set_ylabel('prediction distribution')
+        ax.set_xlabel(r'$P({{risk>High}}|z_{{{}}})$'.format(attr))
+    if iteration:
+        fig.text(1.0, 0.9, f"Training iteration #{iteration}", fontsize='16')
+    if val_metrics is not None:
+        fig.text(1.0, 0.65, '\n'.join(["Prediction performance:",
+                                       f"- ROC AUC: {val_metrics['ROC AUC']:.2f}",
+                                       f"- Accuracy: {val_metrics['Accuracy']:.1f}"]),
+                 fontsize='16')
+    if p_rules is not None:
```
Comment (Mitchellpkt) [inline src/analysis/visualize.py:39]: `is not None` gets the job done, but the double negative seems odd. Is there a more straightforward way?

File: src/model/classifier.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    def __init__(self):
+        pass
+
+    """
+    Create a NN classifier given the shape number of features as input tensor.
+    Returns the created model
+    """
+    def create_nn_classifier(self,n_features):
+        inputs = Input(shape=(n_features,))
+        dense1 = Dense(32, activation='relu')(inputs)
```
Comment (Mitchellpkt) [inline src/model/classifier.py:18]: Where did hard-coded `32` come from?

File: src/model/classifier.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        dropout1 = Dropout(0.2)(dense1)
```
Comment (Mitchellpkt) [inline src/model/classifier.py:19]: Where did hard-coded `2` come from?

File: src/notebooks/main.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import sys
+import pandas as pd
+from sklearn.metrics import accuracy_score, roc_auc_score
+
+sys.path.append('../preprocess/')
```
Comment (Mitchellpkt) [inline src/notebooks/main.py]: Kudos for using relative directories (`../`) however the hard-coded file separator `/` will only work on Linux and Macs. Your code will crash on Windows computers, which use a `\` fileseparator.

Instead, use `os.path.join`

File: src/preprocess/preprocess.py [hunk 8]
```diff
@@ -32,7 +36,7 @@ def loadProcessData(path):
     """
     Split the data into train and test set.
     """
-    def split_data(X, y, Z):
+    def split_data(self,X, y, Z):
         X_train, X_test, y_train, y_test, Z_train, Z_test = train_test_split(X, y, Z, test_size=0.3, stratify=y,
                                                                              random_state=7)
         # standerize the data
```
Comment (Mitchellpkt) [inline src/preprocess/preprocess.py]: (typo)

**Review by Mitchellpkt — CHANGES_REQUESTED** [review]
Great work!

Several of my comments are stylistic, and thus optional depending on your personal preferences.

Several pertain to hard-coded constants. Unless there is a particular reason to manually code in an integer, I prefer to have it dynamically determined (so that the code is extremely flexible for reuse). This can be lower priority - if you're focused on getting new functionality working now, you can wrap back around to cleaning up constants later.

The one thing that must be fixed is the non-universal file separator (since potential future employers may be running your code on their Windows work machines, you definitely want to ensure compatibility). 

I marked "request changes." As soon as you fix the file separator issue, I will consider this good to merge. :- ) The style/constants are lower priority and I will not delay this PR for them 👍

File: src/notebooks/main.py (hunk 7 above)
Comment (suraz09) [reply src/notebooks/main.py]: I agree. Fixed this issue in the new commit for this PR

File: src/analysis/visualize.py (hunk 4 above)
Comment (suraz09) [reply src/analysis/visualize.py:39]: Could not find other alternatives. Still looking into it though, will surely keep an eye on this

File: src/analysis/visualize.py (hunk 2 above)
Comment (suraz09) [reply src/analysis/visualize.py]: Added some comments and lumped up the imports

File: src/analysis/visualize.py (hunk 1 above)
Comment (suraz09) [reply src/analysis/visualize.py:2]: Lumped up all together

File: src/analysis/visualize.py (hunk 3 above)
Comment (suraz09) [reply src/analysis/visualize.py:26]: Will surely look into these hard-coded value and fix them in the next PR

File: src/model/classifier.py (hunk 6 above)
Comment (suraz09) [reply src/model/classifier.py:19]: Will remove all these hard-coded value

**Comment by Mitchellpkt** [comment]
Hi @suraz09 these fixes look great, thanks! Approved! :- )
